src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +789,76 @@ private void PopulateEncLogTableParameters()
             }
         }
 
+        private void PopulateEncLogTableCustomAttributes()
+        {
+            var attributeMap = CreateExistingAttributeMap(_previousGeneration.MetadataReader, _previousGeneration.CustomAttributesAdded, out var lastRow);
+
+            var nextCustomAttributeRow = lastRow + 1;
+
+            foreach (var customAttributeTarget in _customAttributeTargets)
+            {
+                int row;
+                if (attributeMap.TryGetValue(customAttributeTarget, out var queue) &&
+                    queue.Count > 0)
+                {
+                    // If there is already an attribute for this parent, then pull the next available row number
+                    // off the queue and update it
+                    row = queue.Dequeue();
+                }
+                else
+                {
+                    // Otherwise, either this is a new parent that hasn't had an attribute before, or we've run
+                    // out of existing rows to update. Either way, we want to add to the end of the table.
+                    row = nextCustomAttributeRow++;
+
+                    // Keep track of the addition for passing to the next generation
+                    _customAttributesAdded[row] = customAttributeTarget;
+                }
+
+                _customAttributeRows.Add(row);
+                metadata.AddEncLogEntry(
+                    entity: MetadataTokens.CustomAttributeHandle(row),
+                    code: EditAndContinueOperation.Default);
+            }
+        }
+
+        private static Dictionary<EntityHandle, Queue<int>> CreateExistingAttributeMap(MetadataReader metadataReader, IReadOnlyDictionary<int, EntityHandle> customAttributesAdded, out int lastRow)
```

Comment:
>CreateExistingAttributeMap [](http://example.com/codeflow?start=60&length=26)

Let's also handle the case of removing CAs - for now we can emit them with parent field zeroed.

Reply:
Zeroing the parent doesn't work, the runtime throws: https://source.dot.net/#System.Reflection.Metadata/System/Reflection/Metadata/Ecma335/CodedIndex.cs,270

We could leave deleting as a rude edit for now?

Reply:
Ah, right. The spec doesn't allow this.